File: ios/Classes/renderer/CustomRender.swift
import Flutter
import OpenGLES.ES3
import GLKit

@objc public class CustomRender: NSObject, FlutterTexture {
  var fboTargetPixelBuffer: CVPixelBuffer?;
  var fboTextureCache: CVOpenGLESTextureCache?;
  var fboTexture: CVOpenGLESTexture? = nil;
  var fboId: GLuint = 0;

  var glWidth: Double = 640;
  var glHeight: Double = 480;

  //var eAGLShareContext: EAGLContext?;
  var eglEnv: EglEnv?;
  var shareEglCtx: EAGLContext?;

  var worker: RenderWorker? = nil;

  var disposed: Bool = false;

  init(shareContext: EAGLContext?) {
    self.shareEglCtx = shareContext;
    
    super.init();

    //self.eAGLShareContext = EAGLContext.init(api: EAGLRenderingAPI.openGLES3);
    
    self.setup();
  }
  
  func setup() {
    initEGL();
    self.worker = RenderWorker();
    self.worker!.setup();
  }
  
  func getEgl() -> Array<Int64> {
    var _egls = [Int64](repeating: 0, count: 6);
    _egls[2] = self.eglEnv!.getContext();
    return _egls;
  }
  
  func updateTexture(sourceTexture: Int64) -> Bool {
    glEnable(GLenum(GL_BLEND));
    glBlendFunc(GLenum(GL_SRC_ALPHA), GLenum(GL_ONE_MINUS_SRC_ALPHA));
 
    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), fboId);

    glClear(GLbitfield(GL_COLOR_BUFFER_BIT) | GLbitfield(GL_DEPTH_BUFFER_BIT) | GLbitfield(GL_STENCIL_BUFFER_BIT));

    self.worker!.renderTexture(texture: GLuint(sourceTexture), matrix: nil, isFBO: false);

    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), 0);

    glFinish();

    // TODO: callback to flutter notifying texture updated
    // flutter side can do texture copy immediately to its local context
    // then use the local texture copy for further rendering
    
    return true;
  }
  
  public func copyPixelBuffer() -> Unmanaged<CVPixelBuffer>? {
    var pixelBuffer: CVPixelBuffer? = nil;
    pixelBuffer = fboTargetPixelBuffer;
    if (pixelBuffer != nil) {
      let result = Unmanaged.passRetained(pixelBuffer!);
      return result;
    } else {
      print("pixelBuffer is nil.... ");
      return nil;
    }
  }
  
  // ==================================
  func initEGL() {
    self.eglEnv = EglEnv();
    self.eglEnv!.setupRender(shareContext: shareEglCtx);
    self.eglEnv!.makeCurrent();

    initOffscreenFBO(context: self.eglEnv!.context!);
  }

  func initOffscreenFBO(context: EAGLContext) {
    print("FlutterGL initGL  glWidth \(glWidth) glHeight: \(glHeight)  screenScale: \(screenScale)  ");

    self.createCVBufferWithSize(
      size: CGSize(width: glWidth, height: glHeight),
      context: context
    );
    
    checkGlError(op: "EglEnv initGL 11...")

    if(glCheckFramebufferStatus(GLenum(GL_FRAMEBUFFER)) != GL_FRAMEBUFFER_COMPLETE) {
      print("failed to make complete framebuffer object \(glCheckFramebufferStatus(GLenum(GL_FRAMEBUFFER)))");
    }
    
    glBindTexture(CVOpenGLESTextureGetTarget(fboTexture!), CVOpenGLESTextureGetName(fboTexture!));
      
    checkGlError(op: "EglEnv initGL 2...")

    glEnable(GLenum(GL_BLEND));
    glBlendFunc(GLenum(GL_ONE), GLenum(GL_ONE_MINUS_SRC_ALPHA));
    
    glEnable(GLenum(GL_CULL_FACE));
    
    glViewport(0, 0, GLsizei(glWidth), GLsizei(glHeight));
    
    checkGlError(op: "EglEnv initGL 1...")

    var colorRenderBuffer: GLuint = GLuint();
    
    glGenRenderbuffers(1, &colorRenderBuffer);
    glBindRenderbuffer(GLenum(GL_RENDERBUFFER), colorRenderBuffer);
    
    glRenderbufferStorage(GLenum(GL_RENDERBUFFER), GLenum(GL_DEPTH24_STENCIL8), GLsizei(glWidth), GLsizei(glHeight));
    
    glGenFramebuffers(1, &fboId);
    glBindFramebuffer(GLenum(GL_FRAMEBUFFER), fboId);
    glFramebufferTexture2D(GLenum(GL_FRAMEBUFFER), GLenum(GL_COLOR_ATTACHMENT0), GLenum(GL_TEXTURE_2D), CVOpenGLESTextureGetName(fboTexture!), 0);
    glFramebufferRenderbuffer(GLenum(GL_FRAMEBUFFER), GLenum(GL_DEPTH_ATTACHMENT), GLenum(GL_RENDERBUFFER), colorRenderBuffer);
    
    glFramebufferRenderbuffer(GLenum(GL_FRAMEBUFFER), GLenum(GL_STENCIL_ATTACHMENT), GLenum(GL_RENDERBUFFER), colorRenderBuffer);
    
    if(glCheckFramebufferStatus(GLenum(GL_FRAMEBUFFER)) != GL_FRAMEBUFFER_COMPLETE) {
      print("failed to make complete framebuffer object \(glCheckFramebufferStatus(GLenum(GL_FRAMEBUFFER)))");
    }
    
    checkGlError(op: "EglEnv initGL 2...")
  }
  
  func createCVBufferWithSize(size: CGSize, context: EAGLContext) {
    let err: CVReturn = CVOpenGLESTextureCacheCreate(kCFAllocatorDefault, nil, context, nil, &textureCache);
      
    let attrs = [
      kCVPixelBufferPixelFormatTypeKey: NSNumber(value: kCVPixelFormatType_32BGRA),
      kCVPixelBufferOpenGLCompatibilityKey: kCFBooleanTrue,
      kCVPixelBufferOpenGLESCompatibilityKey: kCFBooleanTrue,
      kCVPixelBufferIOSurfacePropertiesKey: [:]
    ] as CFDictionary
    
    let cv2: CVReturn = CVPixelBufferCreate(kCFAllocatorDefault, Int(size.width), Int(size.height),
                                            kCVPixelFormatType_32BGRA, attrs, &fboTargetPixelBuffer);
       
    
    let cvr: CVReturn = CVOpenGLESTextureCacheCreateTextureFromImage(kCFAllocatorDefault,
                                                                     fboTextureCache!,
                                                                     fboTargetPixelBuffer!,
                                                                     nil,
                                                                     GLenum(GL_TEXTURE_2D),
                                                                     GL_RGBA,
                                                                     GLsizei(size.width),
                                                                     GLsizei(size.height),
                                                                     GLenum(GL_BGRA),
                                                                     GLenum(GL_UNSIGNED_BYTE),
                                                                     0,
                                                                     &fboTexture);   
  }

  func checkGlError(op: String) {
    let error = glGetError();
    if (error != GL_NO_ERROR) {
      print("ES30_ERROR", "\(op): glError \(error)")
    }
  }

  func dispose() {
    self.disposed = true;

    //self.eAGLShareContext = nil;
    
    self.eglEnv!.dispose();
    self.eglEnv = nil;
    
    EAGLContext.setCurrent(nil);
  }
  
}

File: android/src/main/kotlin/com/richardyang/better_player/renderer/CustomRender.kt
package com.richardyang.better_player

import android.content.Context
import android.graphics.SurfaceTexture
import android.opengl.GLES32.*
import android.opengl.*
import android.os.Handler
import android.os.HandlerThread
import java.util.concurrent.Semaphore
import io.flutter.view.TextureRegistry.SurfaceTextureEntry
import android.util.Log

class CustomRender : SurfaceTexture.OnFrameAvailableListener{

    var disposed = false;

    lateinit var worker: RenderWorker
    var options: Map<String, Any>;
    var width: Int;
    var height: Int;

    var glWidth: Int = 1;
    var glHeight: Int = 1;

    var dstSurfaceTexture: SurfaceTexture;
    var dstTextureId: Int;
    var screenScale: Double = 1.0;

    lateinit var eglEnv: EglEnv;

    var maxTextureSize = 4096;

    var renderThread: HandlerThread = HandlerThread("VideoCustomRender");
    var renderHandler : Handler

    lateinit var srcSurfaceTex: SurfaceTexture
    var oesTextureMatrix: FloatArray = FloatArray(4 * 4)
    var sharedEglCtx: EGLContext = EGL14.EGL_NO_CONTEXT

    constructor(options: Map<String, Any>, 
        destTexture: SurfaceTextureEntry, 
        shareContext: EGLContext = EGL14.EGL_NO_CONTEXT) {

        this.options = options;
        this.width = 1280;//options["width"] as Int;
        this.height = 720;//options["height"] as Int;
        this.screenScale = 1.0;//options["dpr"] as Double;

        this.dstSurfaceTexture = destTexture.surfaceTexture();
        this.dstTextureId = destTexture.id().toInt();

        this.sharedEglCtx = shareContext

        renderThread.start()
        renderHandler = Handler(renderThread.looper)

        this.executeSync {
            setup();
        }
    }

    fun setup() {
        glWidth = (width * screenScale).toInt()
        glHeight = (height * screenScale).toInt()

        this.initEGL();

        this.worker = RenderWorker();
        this.worker.setup();

        srcSurfaceTex = SurfaceTexture(this.worker!!.srcTextureId)
        srcSurfaceTex.setOnFrameAvailableListener(this, renderHandler)
    }

    fun updateTextureSize(w: Int, h: Int) {
        this.executeSync {
            this.worker.updateTextureSize(w, h)
        }
    }

    override fun onFrameAvailable(videoTexture: SurfaceTexture): Unit {
        glEnable(GL_BLEND);
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
        
        glActiveTexture(GL_TEXTURE1)
        videoTexture.updateTexImage()
        videoTexture.getTransformMatrix(oesTextureMatrix)

        this.worker.renderTexture(oesTextureMatrix);
        //this.worker.renderTexture(oesTextureMatrix, true);

        glFinish();

        checkGlError("update texture 01");
        eglEnv.swapBuffers();

        // TODO: callback to flutter notifying texture updated
        // flutter side can do texture copy immediately to its local context
        // then use the local texture copy for further rendering
    }

    fun initEGL() {
        eglEnv = EglEnv();
        eglEnv.setupRender(this.sharedEglCtx);
        eglEnv.buildWindowSurface(dstSurfaceTexture);
        eglEnv.makeCurrent();
    }

    fun executeSync(task: () -> Unit) {
        val semaphore = Semaphore(0)
        renderHandler.post {
            task.invoke()
            semaphore.release()
        }
        semaphore.acquire()
    }

    fun execute(task: () -> Unit) {
        renderHandler.post {
            task.invoke()
        }
    }

    fun getEgl() : List<Long> {
        var _res = mutableListOf<Long>();
        _res.addAll(this.eglEnv.getEgl());
        return _res;
    }

    fun dispose() {
        disposed = true;
        this.eglEnv.dispose();
        this.worker.dispose();
    }

    fun checkGlError(op: String) {
        val error: Int = glGetError();
        if (error != GL_NO_ERROR) {
            println("ES20_ERROR ${op}: glError ${error}")
            throw RuntimeException("$op: glError $error")
        }
    }
}
